### DISCUSSION TITLE
"This result object does not return rows" on `.one_or_none` on `update` with `execution_options(synchronize_session="fetch")`

### DISCUSSION THREAD

**Discussion by SamuelMarks** [sqlalchemy/sqlalchemy#5940] [Usage Questions]
**Describe the bug**
Error on `.one_or_none` when only one or no element should be provided

**Expected behavior**
Maybe I'm just using the new 1.4/2.0 syntax incorrectly, what I want to do is update an entity and return it, unless the entity doesn't exist in which case return `None`.

**To Reproduce**

```py
from sqlalchemy import Column, String, create_engine, update
from sqlalchemy.orm import Session, declarative_base

engine = create_engine(
    "sqlite://",
    echo=True,
    future=True,
)

Base = declarative_base()


class Config(Base):
    __tablename__ = "config_tbl"

    dataset_name = Column(
        String, doc="name of dataset", default="mnist", primary_key=True
    )


Base.metadata.create_all(engine)


with Session(engine) as session:
    stmt = update(Config) \
                 .where(Config.dataset_name == "wow") \
                 .values({"dataset_name": "baz"}) \
                 .execution_options(synchronize_session="fetch")

config = session.execute(stmt).one_or_none()
```

**Error**

```
Traceback (most recent call last):
  File "delme.py", line 27, in <module>
    config = session.execute(stmt).one_or_none()
  File "lib/python3.8/site-packages/sqlalchemy/engine/result.py", line 1093, in one_or_none
    return self._only_one_row(True, False, False)
  File "lib/python3.8/site-packages/sqlalchemy/engine/result.py", line 559, in _only_one_row
    row = onerow(hard_close=True)
  File "lib/python3.8/site-packages/sqlalchemy/engine/cursor.py", line 1733, in _fetchone_impl
    return self.cursor_strategy.fetchone(self, self.cursor, hard_close)
  File "lib/python3.8/site-packages/sqlalchemy/engine/cursor.py", line 863, in fetchone
    return self._non_result(result, None)
  File "lib/python3.8/site-packages/sqlalchemy/engine/cursor.py", line 914, in _non_result
    result._metadata._we_dont_return_rows(err)
  File "lib/python3.8/site-packages/sqlalchemy/engine/cursor.py", line 1178, in _we_dont_return_rows
    util.raise_(
  File "lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 198, in raise_
    raise exception
sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.
```

**Versions.**
 - OS: macOS Big Sur 11.1 (20C69)
 - Python: 3.8.7
 - SQLAlchemy: 1.4.0b2
 - Database: SQLite (in memory)
 - DBAPI:

**Comment by zzzeek** [comment]
SQL UPDATE statements don't return rows unless they make use of "RETURNING".    The "RETURNING" construct is available with SQLAlchemy using the [returning method](https://docs.sqlalchemy.org//en/14/core/dml.html?highlight=update%20returning#sqlalchemy.sql.expression.Update.returning) - however, SQLite does not support RETURNING.  Additionally, the ORM does not integrate the feature of loading full ORM objects in response to rows yielded by RETURNING (**edit:**: this is now part of 1.4.0b3).  The best you could do would be to get the list of primary key values back and then retrieve that series of primary keys using a SELECT statement against those primary keys.    Or more simply, just loading the objects you want first, then updating them using normal unit of work procedures.

In your specific example, you're using an option called "synchronize_session='fetch'",which does in fact take the steps needed to identify the primary key values that are matched by this statement, which depending on backend may be using RETURNING or may be using an additional SELECT statement; on SQLite, it's going to use a SELECT statement that is in addition to the UPDATE statement emitted.   However, this RETURNING/SELECT internally is used as mentioned above just to locate primary key values, and from there, it does not actually load any ORM objects; it only identifies those objects that are *already loaded* and present in the local `Sesson()` so that it can refresh the attributes that were updated.

Put another way, "synchronize_session" only refers to the strategy to take in order to accommodate the task of synchronizing in-memory state with what the SQL statement has performed.  That is, if you already had some Config objects loaded, and some of them were impacted by your UPDATE statement, they'd not have stale data in them after the "synchronize" operation took place.

From there we can try to address what it is you want to do which is load some entities and updating them, so the most straightforward way would be to emit a select() for those entities, get them back, then update the objects as needed using the ordinary unit of work flush() process.   Especially since your statement targets only one object, this would be the appropriate operation:


```python
with Session(engine) as session:
    stmt = select(Config) \
                 .where(Config.dataset_name == "wow") 
    config = session.execute(stmt).one_or_none()
    if config is not None:
        config.dataset_name = "baz"
    session.commit()



```

**Reply by SamuelMarks** [reply]
Wouldn't there be an atomicity issue with this approach? - I.e., `SELECT` the row, update it, then `SELECT` again? - Or update then select? - In either scenario the row could be deleted…

```py
dataset_name = "wow"
config = Config(dataset_name=dataset_name)

with Session(engine) as session:
    stmt = update(Config) \
                 .where(Config.dataset_name == dataset_name) \
                 .values({"dataset_name": "baz"}) \
                 .execution_options(synchronize_session="fetch")
    config = session.execute(select(Config) \
                 .where(Config.dataset_name == dataset_name)).one_or_none()
```

**Reply by zzzeek** [reply]
There is automatic optimistic concurrency control in SQLAlchemy's unit of work for the "row was deleted" use case, in the common case that the transaction isolation level is not serializable.     the UPDATE emitted by the unit of work does a rowcount to ensure that the number of rows expected, based on primary key, were actually matched; if not, an error is raised.

As you are using a SQLite in-memory database, none of this apples in any case as there's no real concurrency supported in that case.   SQLite does not offer any means of RETURNING for UPDATE in any case so if you are looking to run an UPDATE and get rows back at the same time, you need to use a different database backend.  

SQLite when using a file-based backend is also not appropriate for highly concurrent applications as it relies upon file locking, and the pysqlite driver itself is also a little bit difficult to predict as to when it allows this locking to take place.

now, if your use of SQLite here is just for testing, and you are actually planning to target a different backend, then you can mention what that would be and we can look into what that backend offers in this regard.

**Reply by SamuelMarks** [reply]
Yes SQLite is just for testing. My usual goto is Postgresql, in this scenario however I'm writing a framework so the users of the framework could pick a different engine…

**Reply by zzzeek** [reply]
it's going to be very difficult to create a framework that can make concurrency guarantees in a database agnostic way.

**Reply by zzzeek** [reply]
so the immediate option for PG is to to SELECT...FOR UPDATE which will place a lock on the row, then the unit of work emits the UPDATE.

I tried to see if we can feed an update().returning() right into the ORM, and it seems like it is disallowed at the moment, however, this *should* be possible:

```
    stmt = update(A).filter_by(data="Asdf").returning(A)

    stmt = select(A).from_statement(stmt).execution_options(populate_existing=True)

    a1 = s.execute(stmt).scalar_one()

```

there's no reason that shouldn't be possible by now so let me see if that works if add update/delete with returning to what's accepted there.

**Reply by zzzeek** [reply]
I'm going to push a commit that will allow this to work directly but I think you could also get it right now using a SELECT that selects from an UPDATE-based CTE for PostgreSQL.   This is getting pretty esoteric though so let me push up the new feature and there will be some docs for it too.
